npm package support

muzam1l opened the pull request.

Commit:
pack support

vvp commented on package.json:
```diff
@@ -3,8 +3,7 @@
   "scripts": {
-    "preinstall": "npx only-allow pnpm",
-    "postinstall": "pnpm run generate:typechain",
+    "prepare": "pnpm run compile:sol && pnpm run generate:typechain",
```
I'd like to keep the only-allow check, is it feasible to include in the prepare-phase?

muzam1l replied:
Actually it forces the package using contracts to use pnpm. Other scripts already explicitly use pnpm, so shouldn't be a problem

muzam1l replied:
For example it would force sdk to use pnpm, since that uses contracts (in my pr)

muzam1l replied:
But yeah, maybe we can use it in prepare, I overlooked that part 😅.

vvp replied:
Ok got it, so let's leave it out then :+1: 

And it's not a big problem, we pretty clearly document our pnpm usage anyway in README.md - I'm just not 100% sure what happens to `node_modules` and `package-lock.json` if you install first with pnpm and then npm :thinking:

muzam1l replied:
Oh,  if we want to force pnpm for installing deps, then even `prepare` won't help. Unlike mixing yarn/npm where they show warnings, there is no such warning in case of pnpm. Problem with mixing different package managers is missing dependencies, so those errors will show in build step.

Commit:
Update package name

Commit:
Add typechain definitions to repo

Commit:
Specify the package as a module

Commit:
Extend prettier config

Commit:
Update entrypoint location

muzam1l commented on package.json:
```diff
@@ -1,10 +1,11 @@
   "scripts": {
-    "preinstall": "npx only-allow pnpm",
-    "postinstall": "pnpm run generate:typechain",
+    "postinstall": "npm run compile:sol & npm run generate:typechain",
```
Shouldn't this be `prepare` or something instead of `postinstall`. `postinstall` also runs when some other packages (`sdk`) installs it via `npm install ...`

JaniAnttonen replied:
I think postinstall is actually the best way to generate the typechain definitions, if it works without requiring a pnpm install

Commit:
Move away from pnpm to yarn

Commit:
Update packages

Commit:
Update tests to fit updated packages

Commit:
Publish hardhat config files for typechain generation

Commit:
Move deps from dev-deps to deps

Commit:
Changed globs to find abis in parent library and added test/ in pack.

Skip generating test types on postinstall

Don't publish typechain defs as they're generated on postinstall

Commit:
compile artifacts before packing and generate typechains on postinstall.

muzam1l commented:
It is coming together nicely now, artifacts are compiled before packing and typechains are then generated on postinstall now.

Commit:
Publish config to trick compile on postinstall

Commit:
Trade DX for UX, update Readme

Commit:
Let's try this out

Commit:
INIT_CWD is magic

Commit:
Is PWD more magic?

Commit:
Try some more magic

Commit:
Ditch the postinstall, just bundle the stuff

Commit:
Cross env shell for reading variables
